### PR TITLE
[Merged by Bors] - feat (AlgebraicGeometry/Modules): stalks of the sheaf `M^~`

### PR DESCRIPTION
**Pull request by jjaassoonn**
Construct the isomorphism of $R$-modules from the stalk of the sheaf $M^{\~}$ on $Spec(R)$ to the localization $M_x$ as an $R$-module, where $R$ is a commutative ring and $M$ is an $R$-module. 

## Main definition

* `ModuleCat.tilde.stalkIso`: The isomorphism of `R`-modules from the stalk of `M^~` at `x` to
the localization of `M` at the prime ideal corresponding to `x`.

## Technical note

To get the `R`-module structure on the stalks on `M^~`, we had to define
`ModuleCat.tildeInModuleCat`, which is `M^~` seen as sheaf of `R`-modules. We get it by
applying a forgetful functor to `ModuleCat.tilde M`.

Co-authored-by: Weihong Xu <weihong@vt.edu>
Co-authored-by: Sophie Morel <sophie.morel@ens-lyon.fr>
Co-authored-by: Amelia Livingston <al3717@ic.ac.uk>

This contribution was created as part of the AIM workshop "Formalizing algebraic geometry" in
June 2024.

---

- [x] depends on: #14254
- [x] depends on: #14231
- [x] depends on: #14809 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by jjaassoonn** [comment]
Should be compiling; sorry free; but need clean up

**Comment by riccardobrasca** [comment]
Is this ready for review? If not, can you add "WIP"? Thanks!

**Comment by smorel394** [comment]
Hi Jujian, I did not realize that you had already PRed this branch! Anyway I tried to clean it up, put all the `M^~` in the same file in `AlgebraicGeometry/Modules`, added doc strings etc. If I messed something up, feel free to revert to an earlier commit.
Also, I got a strange error from CI last time I tried to build, it's complaining that two lemmas are not in `simp` normal form, but the expression it wants me to put on the LHS looks identical to the expression that is already there, so I had no idea what was going on and I gave up.

**Comment by jjaassoonn** [comment]
> Hi Jujian, I did not realize that you had already PRed this branch! Anyway I tried to clean it up, put all the `M^~` in the same file in `AlgebraicGeometry/Modules`, added doc strings etc. If I messed something up, feel free to revert to an earlier commit. Also, I got a strange error from CI last time I tried to build, it's complaining that two lemmas are not in `simp` normal form, but the expression it wants me to put on the LHS looks identical to the expression that is already there, so I had no idea what was going on and I gave up. As far as I can tell, this is the last remaining CI problem.


Thanks for doing this! I have no idea either, so I just removed the `simp` attribute

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 1]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+  Limits.colimit.ι_desc _ _
+
+@[simp]
+theorem stalkToFiberLinearMap_germ (U : Opens (PrimeSpectrum.Top R)) (x : PrimeSpectrum.Top R)
+    (hx : x ∈ U) (s : (tildeInModuleCat M).1.obj (op U)) :
+    stalkToFiberLinearMap M x
+      (TopCat.Presheaf.germ (tildeInModuleCat M) U x hx s) = (s.1 ⟨x, hx⟩ : _) :=
+  DFunLike.ext_iff.1 (germ_comp_stalkToFiberLinearMap M U x hx) s
+
+@[simp]
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:293]:
```suggestion
@[reassoc (attr := simp)]
```

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 2]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+theorem toOpen_germ (U : Opens (PrimeSpectrum.Top R)) (x) (hx : x ∈ U) :
+    toOpen M U ≫ TopCat.Presheaf.germ (tildeInModuleCat M) U x hx = toStalk M x := by
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:295]:
```suggestion
    toOpen M U ≫ M.tildeInModuleCat.germ U x hx = toStalk M x := by
```

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 3]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+  rw [← toOpen_germ]; rfl
+
+@[simp]
+theorem toStalk_comp_stalkToFiberLinearMap (x : PrimeSpectrum.Top R) :
+    toStalk M x ≫ stalkToFiberLinearMap M x =
+    LocalizedModule.mkLinearMap x.asIdeal.primeCompl M := by
+  rw [toStalk, Category.assoc, germ_comp_stalkToFiberLinearMap]; rfl
+
+theorem stalkToFiberLinearMap_toStalk (x : PrimeSpectrum.Top R) (m : M) :
+    (stalkToFiberLinearMap M x) ((toStalk M x) m) =
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:309]:
```suggestion
    (stalkToFiberLinearMap M x) (toStalk M x m) =
```

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 4]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+  hom_inv_id := TopCat.Presheaf.stalk_hom_ext _ fun U hxU ↦ ext _ fun s ↦ by
+    show localizationToStalk M x (stalkToFiberLinearMap M x (M.tildeInModuleCat.germ U x hxU s)) =
+      M.tildeInModuleCat.germ U x hxU s
+    rw [stalkToFiberLinearMap_germ]
+    obtain ⟨V, hxV, iVU, f, g, (hg : V ≤ PrimeSpectrum.basicOpen _), hs⟩ :=
+      exists_const _ _ s x hxU
+    rw [← res_apply M U V iVU s ⟨x, hxV⟩, ← hs, const_apply, localizationToStalk_mk]
+    exact (tildeInModuleCat M).germ_ext V hxV (homOfLE hg) iVU <| hs ▸ rfl
+  inv_hom_id := by ext x; exact x.induction_on (fun _ _ => by simp)
+
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:390]: Can you provide an `IsLocalizedModule` instance?

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 5]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+    TopCat.Presheaf.stalk (tildeInModuleCat M) x ⟶
+    ModuleCat.of R (LocalizedModule x.asIdeal.primeCompl M) :=
+  Limits.colimit.desc ((OpenNhds.inclusion x).op ⋙ (tildeInModuleCat M))
+    { pt := _
+      ι :=
+      { app := fun U => openToLocalization M ((OpenNhds.inclusion _).obj U.unop) x U.unop.2 } }
+
+@[simp]
+theorem germ_comp_stalkToFiberLinearMap (U : Opens (PrimeSpectrum.Top R)) (x) (hx : x ∈ U) :
+    TopCat.Presheaf.germ (tildeInModuleCat M) U x hx ≫ stalkToFiberLinearMap M x =
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:282]:
```suggestion
    (tildeInModuleCat M).germ U x hx ≫ stalkToFiberLinearMap M x =
```

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 6]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+  toFun s := (s.1 ⟨x, hx⟩ : _)
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+/--
+The morphism of `R`-modules from the stalk of `M^~` at `x` to the localization of `M` at the
+prime ideal of `R` corresponding to `x`.
+-/
+noncomputable def stalkToFiberLinearMap (x : PrimeSpectrum.Top R) :
+    TopCat.Presheaf.stalk (tildeInModuleCat M) x ⟶
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:273]:
```suggestion
    (tildeInModuleCat M).stalk  x ⟶
```

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 7]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
     (TopCat.Presheaf.stalk (tildeInModuleCat M) x) :=
   LocalizedModule.lift _ (toStalk M x) <| isUnit_toStalk M x
 
+
+/-- The ring homomorphism that takes a section of the structure sheaf of `R` on the open set `U`,
+implemented as a subtype of dependent functions to localizations at prime ideals, and evaluates
+the section on the point corresponding to a given prime ideal. -/
+def openToLocalization (U : Opens (PrimeSpectrum R)) (x : PrimeSpectrum R) (hx : x ∈ U) :
+    (tildeInModuleCat M).1.obj (op U) ⟶
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:262]:
```suggestion
    (tildeInModuleCat M).presheaf.obj (op U) ⟶
```

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 8]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+      (TopCat.Presheaf.germ (tildeInModuleCat M) U x hx s) = (s.1 ⟨x, hx⟩ : _) :=
+  DFunLike.ext_iff.1 (germ_comp_stalkToFiberLinearMap M U x hx) s
+
+@[simp]
+theorem toOpen_germ (U : Opens (PrimeSpectrum.Top R)) (x) (hx : x ∈ U) :
+    toOpen M U ≫ TopCat.Presheaf.germ (tildeInModuleCat M) U x hx = toStalk M x := by
+  rw [← toOpen_res M ⊤ U (homOfLE le_top : U ⟶ ⊤), Category.assoc, Presheaf.germ_res]; rfl
+
+theorem germ_toOpen (U : Opens (PrimeSpectrum.Top R)) (x) (hx : x ∈ U) (f : M) :
+    (M.tildeInModuleCat.germ U x hx) ((ModuleCat.Tilde.toOpen M U) f) = toStalk M x f := by
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:299]:
```suggestion
    M.tildeInModuleCat.germ U x hx (ModuleCat.Tilde.toOpen M U f) = toStalk M x f := by
```

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Modules/Tilde.lean:299]: Does this lemma come for free if you put `@[elementwise]` on `toOpen_germ`?

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 9]
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+  rw [← toOpen_germ]; rfl
+
+@[simp]
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:302]:
```suggestion
@[reassoc (attr := simp)]
```

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean (hunk 3 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Modules/Tilde.lean:309]: Does `elementwise` work here?

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean (hunk 8 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Modules/Tilde.lean]: indeed!

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean (hunk 4 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Modules/Tilde.lean:386]: Thanks. Provided

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 10]
```diff
@@ -252,6 +254,142 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+    show localizationToStalk M x (stalkToFiberLinearMap M x (M.tildeInModuleCat.germ U x hxU s)) =
+      M.tildeInModuleCat.germ U x hxU s
+    rw [stalkToFiberLinearMap_germ]
+    obtain ⟨V, hxV, iVU, f, g, (hg : V ≤ PrimeSpectrum.basicOpen _), hs⟩ :=
+      exists_const _ _ s x hxU
+    rw [← res_apply M U V iVU s ⟨x, hxV⟩, ← hs, const_apply, localizationToStalk_mk]
+    exact (tildeInModuleCat M).germ_ext V hxV (homOfLE hg) iVU <| hs ▸ rfl
+  inv_hom_id := by ext x; exact x.induction_on (fun _ _ => by simp)
+
+instance (x : PrimeSpectrum.Top R) : IsLocalizedModule x.asIdeal.primeCompl (stalkIso M x).hom :=
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:387]: No this means `M_x` is the localization of `M_x` at `x` right?

Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Modules/Tilde.lean]: Indeed

Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Modules/Tilde.lean]: now it should be okay

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 11]
```diff
@@ -252,6 +254,145 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+      M.tildeInModuleCat.germ U x hxU s
+    rw [stalkToFiberLinearMap_germ]
+    obtain ⟨V, hxV, iVU, f, g, (hg : V ≤ PrimeSpectrum.basicOpen _), hs⟩ :=
+      exists_const _ _ s x hxU
+    rw [← res_apply M U V iVU s ⟨x, hxV⟩, ← hs, const_apply, localizationToStalk_mk]
+    exact (tildeInModuleCat M).germ_ext V hxV (homOfLE hg) iVU <| hs ▸ rfl
+  inv_hom_id := by ext x; exact x.induction_on (fun _ _ => by simp)
+
+instance (x : PrimeSpectrum.Top R) :
+    IsLocalizedModule x.asIdeal.primeCompl (toOpen _ _ ≫ M.tildeInModuleCat.germ ⊤ x ⟨⟩) := by
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:388]: The map is just `toStalk` right?

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean [hunk 12]
```diff
@@ -252,6 +254,145 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
...
+    obtain ⟨h1, (h2 : g • s.1 ⟨y, _⟩ = LocalizedModule.mk f 1)⟩ := hfg y
+    exact show LocalizedModule.mk f ⟨g, by exact h1⟩ = s.1 (iVU y) by
+      set x := s.1 (iVU y); change g • x = _ at h2; clear_value x
+      induction x using LocalizedModule.induction_on with
+      | h a b =>
+        rw [LocalizedModule.smul'_mk, LocalizedModule.mk_eq] at h2
+        obtain ⟨c, hc⟩ := h2
+        exact LocalizedModule.mk_eq.mpr ⟨c, by simpa using hc.symm⟩⟩
+
+
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Modules/Tilde.lean:343]:
```suggestion
```

**Review by erdOne — APPROVED** [review]
LGTM otherwise

File: Mathlib/AlgebraicGeometry/Modules/Tilde.lean (hunk 11 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Modules/Tilde.lean]: Indeed. Thanks

**Review by erdOne — APPROVED** [review]
Thanks!
maintainer merge

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**